### PR TITLE
Automate distribution testing

### PR DESCRIPTION
**Pull request by eb8680**
Addresses #386. ~~Blocked by #388.~~

This PR adds a new file `test_distribution_generic.py` that refactors the distribution tests. Here there is one generic test for each distribution method, and all that is needed to test a new distribution is to add a recipe for constructing a random instance to a list of test cases.  This change is necessary if we want to approach full coverage of the dozens of distributions in PyTorch, Pyro/NumPyro and TFP without thousands of lines of manually duplicated testing logic.

I have also had to add a number of small fixes to get the new tests to pass, which I would argue is a sign of the value of these tests - even the first version here is exercising many small distribution API edge cases that would be hard to catch with the current approach.  There are also a couple of new testing utilities (`ops.isnan` and `testing.random_scale_tril`).

If the approach in this PR works, I will delete many of the one-off tests in `test_distribution.py` in a followup PR.

Remaining tasks:
- [x] Get generic gradient test running
- [x] Add enough test cases to reach coverage parity with the tests in `test_distribution.py`
- [x] Get all deterministic tests to pass and avoid nans

Triaged
- ~~Get all Monte Carlo tests to pass~~ - I am skeptical of the viability and usefulness of doing this and have simply disabled and removed these tests for now.

### CODE REVIEW

File: test/test_distribution_generic.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+@pytest.mark.skipif(not _ENABLE_MC_DIST_TESTS, reason="slow and finicky Monte Carlo tests disabled by default")
+@pytest.mark.parametrize("case", TEST_CASES, ids=str)
+@pytest.mark.parametrize("sample_shape", [(200000,), (400, 400)], ids=str)
+@pytest.mark.parametrize("statistic", [
+    "mean",
+    "variance",
+    pytest.param("entropy", marks=[pytest.mark.skipif(get_backend() == "jax", reason="entropy not implemented")])
+])
+def test_generic_stats_sample(case, statistic, sample_shape):
```
Comment (eb8680) [inline test/test_distribution_generic.py:375]: This test comparing Monte Carlo estimates of summary statistics with ground-truth values is slow (especially on the JAX backend) and finicky.  I've disabled it by default but could also remove it entirely - I'm not sure failures are very informative.

File: test/test_distribution_generic.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+@pytest.mark.skipif(not _ENABLE_MC_DIST_TESTS, reason="slow and finicky Monte Carlo tests disabled by default")
+@pytest.mark.skipif(get_backend() != "torch", reason="not working yet on jax")
+@pytest.mark.parametrize("case", TEST_CASES, ids=str)
+@pytest.mark.parametrize("sample_shape", [(200000,), (400, 400)], ids=str)
+@pytest.mark.parametrize("statistic", [
+    "mean",
+    "variance",
+    pytest.param("entropy", marks=[pytest.mark.skipif(get_backend() == "jax", reason="entropy not implemented")])
+])
+def test_generic_grads_sample(case, statistic, sample_shape):
```
Comment (eb8680) [inline test/test_distribution_generic.py:399]: Ditto for this test comparing Monte Carlo estimates of gradients of summary statistics wrt parameters versus gradients of ground-truth values - it's slow and finicky, and I'm not sure failures are very informative.  It's disabled by default, but I am open to removing it entirely.

File: test/test_distribution_generic.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+TEST_CASES = []
+
+for batch_shape in [(), (5,), (2, 3)]:
+
+    # BernoulliLogits
+    TEST_CASES += [DistTestCase(
+        "backend_dist.Bernoulli(logits=case.logits)",
+        (("logits", f"rand({batch_shape})"),),
+        funsor.Real,
+    )]
```
Comment (eb8680) [inline test/test_distribution_generic.py:72]: After this PR, adding a new distribution to `funsor.distributions` will be as simple as adding it to the list of distributions to be wrapped in `funsor/{backend}/distributions.py` and adding a new test case to this list.

File: test/test_distribution_generic.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+        return hash((self.raw_dist, self.raw_params, self.expected_value_domain))
+
+
+TEST_CASES = []
+
+for batch_shape in [(), (5,), (2, 3)]:
+
+    # BernoulliLogits
+    TEST_CASES += [DistTestCase(
+        "backend_dist.Bernoulli(logits=case.logits)",
```
Comment (eb8680) [inline test/test_distribution_generic.py:69]: Note I've chosen to encode test cases with strings of Python code. This should allow us to apply these generic tests even to quite complicated backend distribution expressions, e.g. `TransformedDistribution(Normal(loc, scale).to_event(1), [TanhTransform,]).mask(mask)`

File: test/test_distribution_generic.py (hunk 3 above)
Comment (fritzo) [reply test/test_distribution_generic.py:72]: > adding it to the list

Is there a `.register()` command allowing users to dynamically extend the list in their own code? Similar to the dynamic registration mechanisms in `kl_divergence.register()` or `biject_to.register`? Even better, could we automatically register distributions in Funsor the first time they are encountered?

File: funsor/distribution.py [hunk 5]
```diff
@@ -134,23 +134,28 @@ def unscaled_sample(self, sampled_vars, sample_inputs, rng_key=None):
...
         sample_args = (sample_shape,) if get_backend() == "torch" else (rng_key, sample_shape)
         if self.has_rsample:
-            raw_sample = raw_dist.rsample(*sample_args)
+            raw_value = raw_dist.rsample(*sample_args)
         else:
-            raw_sample = ops.detach(raw_dist.sample(*sample_args))
+            raw_value = ops.detach(raw_dist.sample(*sample_args))
 
-        result = funsor.delta.Delta(value.name, Tensor(raw_sample, inputs, value.output.dtype))
+        funsor_value = to_funsor(raw_value, output=value_output, dim_to_name=dim_to_name)
```
Comment (eb8680) [inline funsor/distribution.py:150]: I had to update `funsor.distribution.Distribution.unscaled_sample` to use `to_funsor` and `to_data` throughout to get some tests to pass.

File: test/test_distribution_generic.py (hunk 3 above)
Comment (eb8680) [reply test/test_distribution_generic.py:72]: > Is there a .register() command allowing users to dynamically extend the list in their own code?

Yes, `funsor.distribution.make_dist` basically plays this role - it takes a backend distribution class and (optionally) some parameter names as input and generates a new `funsor.distribution.Distribution` with generic `eager`, `to_funsor` and `to_data` patterns that should work for most use cases, provided the user has correctly implemented `.arg_constraints` and `.support` in their custom distribution.

**Comment by eb8680** [comment]
@fehiepsi Travis seems to be hanging somewhere in the JAX tests, but I haven't been able to reproduce it locally and I can't figure out where from the logs, although I assume it's related to the distribution tests.  Any idea what might be going on?

**Comment by eb8680** [comment]
One thing I'm noticing is that `test_dirichlet_sample` and the other sampler tests in `test/test_distributions.py` are using huge amounts of memory under the JAX backend both in absolute terms and relative to the PyTorch backend, and may be leaking memory.  This was true before this PR but maybe it got worse?

File: funsor/torch/distributions.py [hunk 6]
```diff
@@ -154,6 +155,19 @@ def _infer_param_domain(cls, name, raw_shape):
...
 
+###########################################################
+# Converting distribution funsors to PyTorch distributions
+###########################################################
+
+@to_data.register(Multinomial)  # noqa: F821
+def multinomial_to_data(funsor_dist, name_to_dim=None):
+    probs = to_data(funsor_dist.probs, name_to_dim)
+    total_count = to_data(funsor_dist.total_count, name_to_dim)
+    if isinstance(total_count, numbers.Number) or len(total_count.shape) == 0:
```
Comment (fritzo) [inline funsor/torch/distributions.py:166]: Do we need to worry about `int(total_count)` thwarting PyTorch tracing? Should we preserve scalar `Tensor`s?

File: test/test_distribution_generic.py [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+class DistTestCase:
+
+    def __init__(self, raw_dist, raw_params, expected_value_domain):
+        self.raw_dist = raw_dist
+        self.raw_params = raw_params
+        self.expected_value_domain = expected_value_domain
+        for name, raw_param in self.raw_params:
+            if get_backend() != "numpy":
+                # we need direct access to these tensors for gradient tests
+                setattr(self, name, eval(raw_param))
```
Comment (fritzo) [inline test/test_distribution_generic.py:54]: Why not simply add `TEST_CASES.append(self)` here, so that creating a `DistTestCase` automatically registers it? IIUC you're simply iterating over all instances, right?

**Review by fritzo — COMMENTED** [review]
...still reviewing test_distribution_generic.py. I think it will be easier to read after adding `TEST_CASES.append(self)` to `.__init__()`.

File: funsor/torch/distributions.py (hunk 6 above)
Comment (eb8680) [reply funsor/torch/distributions.py:166]: I made this change because `torch.distributions.Multinomial` raises an error given a tensor `total_count`, even if it is a scalar tensor, and the generic conversion tests were failing.  It's definitely an issue, but there are already lots of JIT issues with scalars and tuple shapes throughout the codebase (encountered while working on `pyro.contrib.funsor`), and my inclination with this PR and further work on distributions is to avoid special-casing to the greatest extent possible even if that means deferring to odd implementation details in the backends.  I think a better fix in this instance would be to allow scalar tensor `total_count` upstream so this change could be reverted.

This is also a good reminder to add some generic JIT tests for distribution wrappers in a followup PR.

**Comment by fehiepsi** [comment]
@eb8680 I think the memory issue is due to "caching" mechanism of jax. It cached the compiled code so that the next time, a function can be executed fast given inputs with the same shape. I guess we can split
```python
pytest -v -n auto --ignore=test/examples --ignore=test/pyro --ignore=test/pyroapi
```
into smaller procedures, e.g.
```python
pytest -v -n auto --ignore=test/examples --ignore=test/pyro --ignore=test/pyroapi --ignore=test/distribution.py
pytest -v -n auto test/distribution.py
```

**Comment by eb8680** [comment]
Would that also explain the run time differences? For comparison, the new sampler smoke tests in this PR were taking ~1s on a single core on my laptop for PyTorch vs ~100s for JAX

**Comment by fehiepsi** [comment]
How many tests there are? If there are 100 tests, then it is normal to me.

**Comment by eb8680** [comment]
> How many tests there are? If there are 100 tests, then it is normal to me.

Around that, yeah

File: test/test_distribution_generic.py (hunk 1 above)
Comment (eb8680) [reply test/test_distribution_generic.py]: I removed these tests from this PR.

File: test/test_distribution_generic.py (hunk 7 above)
Comment (eb8680) [reply test/test_distribution_generic.py:56]: Done.

**Review by fritzo — APPROVED** [review]
